internal/receiver/awscontainerinsightreceiver/ecs/ecs_util.go

```diff
@@ -0,0 +1,127 @@
+package ecs
+
+import (
+	"fmt"
+	common "github.com/aws-observability/aws-otel-collector/internal/aws/containerinsightcommon"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+	"reflect"
+)
+
+func insertMetrics(key string, value interface{}, ilms pdata.InstrumentationLibraryMetricsSlice, metricType string, timestamp pdata.Timestamp, logger *zap.Logger) {
+
+	metric := common.RemovePrefix(metricType, key)
+	unit := common.GetUnitForMetric(metric)
+
+	switch t := value.(type) {
+	case int:
+		ilms.Append(common.IntGauge(key, unit, int64(t), timestamp))
+	case int32:
+		ilms.Append(common.IntGauge(key, unit, int64(t), timestamp))
+	case int64:
+		ilms.Append(common.IntGauge(key, unit, t, timestamp))
+	case uint:
+		ilms.Append(common.DoubleGauge(key, unit, float64(t), timestamp))
+	case uint32:
+		ilms.Append(common.DoubleGauge(key, unit, float64(t), timestamp))
+	case uint64:
+		ilms.Append(common.DoubleGauge(key, unit, float64(t), timestamp))
+	case float32:
+		ilms.Append(common.DoubleGauge(key, unit, float64(t), timestamp))
+	case float64:
+		ilms.Append(common.DoubleGauge(key, unit, t, timestamp))
+	default:
+		valueType := fmt.Sprintf("%T", value)
+		logger.Warn("Detected unexpected field", zap.String("key", key), zap.Any("value", value), zap.String("value type", valueType))
+	}
+}
+
+func getMetricValue(metricName string, rms pdata.ResourceMetrics) (bool, interface{}, pdata.Timestamp) {
```

Comment:
There is a lot of duplicate code here. See https://github.com/aws-observability/aws-otel-collector/blob/main/internal/processor/awscontainerinsightprocessor/stores/utils.go#L189-L210

As you can see, it is very cumbersome to get a metric value from openTelemetry format. I would suggest to not use it for raw metrics. You can use simple maps to represent the metrics and do processing and then convert them to openTelemetry format at the final stage.